# 异常处理
## 状态恢复机制
Solidity的异常处理机制使用状态恢复的方式。一旦异常产生，将撤销本次调用对于状态变量的更改，同时向调用者返回异常通知；

## 错误类型

* Error

    内置的错误类型之一，有两种类型：`Error()`和`Error(string)`；

    触发场景：

    * 调用reuqire函数且检查条件为false；
    * 使用revert触发（关键字或revert函数）；
    * 执行一个外部函数调用，但目标是一个不含代码的合约；
    * 合约使用非`payable`函数接收Ether；
    * 通过一个公共的getter函数接收Ether；

    >Error有一个错误选择器的属性`error.selector`,可以表明该错误的错误类型；
* Panic

    内置的错误类型之一:`Panic(uint256)`,其中参数为错误代码;

    一个正确运行的程序不应该出现Panic异常，即使是外部输入错误；
    
    [触发场景与错误代码](https://docs.soliditylang.org/zh/v0.8.17/control-structures.html#assert-require-revert-and-exceptions)

* CustomError

    自定义错误类型，定义格式为：`error CustomError(arg1,arg2);`

    自定义错误数据类型是合约的可继承属性，但不能被重写与覆盖；
    
    触发场景:
    
    自定义错误只能在revert语句处触发；

## 触发方式
目前能够触发错误的语句与关键字有三种类型：Assert,Require,Revert;

|  类型   |                 触发形式                  |          错误类型           | 备注                                                                |
| :-----: | :---------------------------------------: | :-------------------------: | :------------------------------------------------------------------ |
| Assert  |             assert(condition)             |       Panic(uint256)        | assert断言适用于内部错误条件检查，一个正确的程序不应该出现Panic错误 |
| Require |        require(condition,string) or require(condition)        |  Error() or Error(string)   | 适用于外部条件检查，返回错误类型于触发形式有关                      |
| Revert  | revert CustomError(arg) | CustomError(arg) |                                                                     |
| Revert  |             revert(string) or revert Error()              |       Error(string) or Error()       |                                                                     |

>使用一个自定义错误通常比使用字符串描述更加便宜（可以使用错误名表述数据），通过使用NatSpec语法可以免费提供更多的注释；


## 捕获与处理

捕获的对象：**外部调用**与**创建新合约**的异常

catch支持的捕获块种类：

```Solidity
try call() returns (uint v) {
    return v;
}catch Error(string memory){

}catch Panic(uint) {

}catch (bytes memroy) {

}catch{
    
}
```

* 第一个捕获块对应内置Error错误，参数string可携带异常信息，用于异常处理；
  
* 第二个捕获块对应内置Panic错误，参数为异常错误代码；

* 第三个捕获块可捕捉任意错误，bytes数据可提供对于低级错误数据的访问；

* 最后一个则默认捕获所有错误，适用于对错误数据不感兴趣的情况；

>在try/catch语句后的参数只可在临近的捕获块中使用；
